### DISCUSSION TITLE
Writing stack config in TypeScript?

### DISCUSSION THREAD

**Discussion by cakoose** [pulumi/pulumi#7290] [Pulumi in Practice] [answered]
I have a TypeScript-based Pulumi setup with two stacks -- prod and staging.  I'm currently putting my stack-specific code in "Pulumi.prod.yaml" and "Pulumi.staging.yaml".

However, I'm thinking that it might be better to write the config in TypeScript instead of YAML.  The advantage would be static types and not having to worry about YAML's parsing gotchas.

Here's roughly what I'm thinking:

```ts
// config.ts
type StackConfig = {
    active: boolean;
    nums: Array<number>;
};

// config_prod.ts and config_staging.ts
const config: StackConfig = { ... }
export default config;

// Then, everywhere I need to do some config...
const config = require(`config_${pulumi.getStack()})`);
if (config.active) {
    ...
}
```

Thoughts on this general approach?  Any downsides I'm missing here?

**Comment by lukehoban** [comment, marked as answer]
It’s definitely reasonable to “compute” your stack configuration values in code from a much smaller set of configuration specified in Pulumi config. A common case for this is project defaults - with currently must be provided in code. But it’s reasonable and fairly common to go beyond that and write config in code and switch on something more narrow from the external config. The only downside is that the codebase is now more constrained to only work for those specific configurations, whereas the standard approach aims to make a codebase  able to support new configuration values that are not yet built in to the codebase without modifying the code itself. But for many usecases, that may not be a distinction that matters much. 

You may also be interested in the feature proposal at https://github.com/pulumi/pulumi/issues/1052 which is related to this.

**Comment by headconnect** [comment]
I'm thinking along the same lines - I've built a highly configurable pulumi program with every single useful resource (for us) defined with sensible defaults.  From this single pulumi program, we create multiple stacks for each environment and customer (dev/stage/prod environments, usually one stack of each environment for each customer, but some customers can have multiple, etc.. ).

Adding to that - this is a three-stage approach, with the baseline infra creating cloud resources, including kubernetes clusters (plural), then a baseline k8s creating the standard kubernetes components (monitoring, choice of databases, security config, etc) - and lastly the application layer deploying on top of that. 

To simplify this and ensuring we can define the "whole customer" - I put all the configuration (from the stack yaml) in the infra outputs, which is read as a stack reference by the others.

There are therefore two conditions where having the stack defined as a .ts export would benefit greatly:

1. The way I've set up the configuration is per-component, such that each component has sensible defaults in code, some explicit complicated settings simplified by boolean or enum configuration (subnet references for instance), and both global and specific full configuration override.

It's the last bit which is most interesting - this technically accepts the same configuration as the provider itself, only `Partial<>`.  Making sure this part of config is valid is tricky when it's all in yaml.  Additionally, since I'm operating on sensible defaults - most of these components accept -zero- configuration, sometimes only requiring the name. Figuring out which other configuration is accessible (and how) requires us to maintain seperate documentation which then has to be up to date with the internal typing of that configuration as used by the resources and component arguments - and requires the users to have both documentation and configuration open.

2. The other benefit is that currently, only the baseline-infra stack consumes that configuration, and it's required that the stack is re-run to accept new configuration changes (which is sane) even though it doesn't affect the infrastructure (cloud resources) side at all.

For instance which databases that customer needs might change, and that's handled in baseline k8s - which could easily just import the same configuration directly so we manage all parts from the same singular configuration. 


Naturally, things like secrets and some very specific aspects would be handled in the stack config yaml - but the vast majority would be better served outside.

My planned strategy would be that we need to run the following:



1. `pulumi stack init .... `
2. `cat stack-ts\{stack-template.ts} > stack-ts\{stack-name.ts} `
3. update `index.ts` in the `stack-ts` folder to export the new `stack-name.ts`
4. In my pulumi program - import the `stack-ts` as an object with string-index for stack-name
5. use `pulumi.getStack()` to get the stack name and load the correct config by `const stackConf = stackTs[pulumi.getStack()]` where stackTs, for instance, is `export const stackTs = { "stack-name": {...conf...}, "another-stack": {...conf} }

```
